Polar tick improvements

QuLogic opened the pull request:
## PR Summary

Semi-WIP as some bits of this are a bit hacky, so they might need to be corrected a bit. This improves upon ticks used in polar plots by correctly rotating them to be perpendicular to the spine. Tick labels also now use the configured padding instead of some fixed value radially; this is a big difference if the radial limits are small or large.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

QuLogic commented:
Would like to get this in to 2.1, since it's a good follow-on to #4699. Just need to figure out why it's not passing somehow.

QuLogic commented:
Ah, all set now; I wonder if that `reset_ticks` change should be moved into the main function?

tacaswell commented:
The r-grid ticks do not look like they are pulling the tangent from the right place?

Over all :+1:

dstansby commented:
I think the new label directions are much more un-readable than having them all horizontal. It also doesn't look like there are any image tests for angular ticks on the outside of the axes?

dopplershift commented:
I'm +0.5 on having the capability to turn on the rotated labels, -1000 on having it be the default. This belonged in 2.0 if we were going to do it since this is a stylistic change--I really don't think our users will view this as a clear enhancement. (My first reaction honestly was "Ewww.")

WeatherGod commented:
Just a quick thought... should we make sure the api for this matches the
api for rotated labels in pie()?

On Tue, Aug 29, 2017 at 12:28 PM, Ryan May <notifications@github.com> wrote:

> I'm +0.5 on having the capability to turn on the rotated labels, -1000 on
> having it be the default. This belonged in 2.0 if we were going to do it
> since this is a stylistic change--I really don't think our users will view
> this as a clear enhancement. (My first reaction honestly was "Ewww.")
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/9068#issuecomment-325719388>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-BWE6RB7Z9LtfUqmmq-ndz8Rm3deks5sdDw-gaJpZM4O933_>
> .
>

efiring commented:
This could be a useful option--but only if the upside-down labels are flipped 180 degrees to bring them right-side-up.  In other words, label orientations should be kept within +-90 degrees, not +-180.

WeatherGod commented:
Just took a look at the rotatelabels feature of pie(), and it made me
realize that there might be two ways people would want the labels oriented:
perpendicular to the normal, or parallel to the normal. Right now, pie()'s
(non-default) rotatelabels feature will orient the labels parallel to the
normal, and also will flip the text such that it is never upside down.

On Tue, Aug 29, 2017 at 12:56 PM, Eric Firing <notifications@github.com>
wrote:

> This could be a useful option--but only if the upside-down labels are
> flipped 180 degrees to bring them right-side-up. In other words, label
> orientations should be kept within +-90 degrees, not +-180.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/9068#issuecomment-325726900>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-ATabNBQFTqrXcVcmz5qCppbcu0eks5sdEKNgaJpZM4O933_>
> .
>

dopplershift commented:
So an [informal survey](https://twitter.com/dopplershift/status/902569901461835776) by posting an image (from the test suite changes) to Twitter would seem to argue that this change is by no means a slam dunk in terms of improving default plot quality. I do think there's a strong case to be made for having the option to turn this behavior on, though.

afvincent commented:
FWIW, even if one decides to keep the current default behavior, *I would advocate for keeping the 2 new polar-tick classes and an ad-hoc example somewhere*.

IMHO, the rotated radial ticks could sometimes be more readable than the current ones, especially when cluttering is not far away, for example (from the test images):
![polar_proj_current](https://user-images.githubusercontent.com/17270724/29849803-0e86a2d8-8cde-11e7-8d35-cef588af2eaa.png)
vs.
![polar_proj_rotated](https://user-images.githubusercontent.com/17270724/29849811-1e16cf5c-8cde-11e7-8149-048b05ff0323.png)
Even though the anchor for rotation seems a bit weird to me, it is rather clear to me which label corresponds to which circle. With the current behavior (and apart from the problem of the radial tick labels overlapping each other...), I challenge anybody to say the same for the intermediate values without counting the circles and the tick labels from one edge or the other 😈.

For the theta-tick label, I think that having an option to limit the rotation domain as @efiring suggested, as well as @WeatherGod´s one about allowing the labels to be parallel to the normal could be nice additions. If I may, I would also suggest to have an option to rotate the theta-tick labels outwards, i.e. + 180° compared to the current rotated ones. Possible uses of rotated theta-tick labels that come to my mind are sky/constellation charts (random examples [here](https://maas.museum/app/uploads/sites/6/2016/05/starmapApril2016-2007x1667.jpeg) and [here](http://www.constellation-guide.com/wp-content/uploads/2014/06/Little-Dipper-and-northern-constellations.jpg))

QuLogic commented:
There is certainly a tradeoff involved between full-circle plots and wedges (and I'm probably favouring the latter a bit). Consider
![figure_1](https://user-images.githubusercontent.com/302469/29853281-fe5eb33a-8d0c-11e7-80d6-1ffa6cbcc043.png)
vs
![figure_2](https://user-images.githubusercontent.com/302469/29853287-0326e612-8d0d-11e7-9523-349bdf3c7d65.png)
The radial ticks could be rotated manually (but it'd be a pain if you modified the limits) but angular ticks would be much more annoying. That being said, there are already several special cases for full-circle plots and having it remain (almost) the same as before is certainly feasible.

> The r-grid ticks do not look like they are pulling the tangent from the right place?

> Even though the anchor for rotation seems a bit weird to me,

They are anchored to fit with ticks. However, since ticks have always been broken in polar plots, they're off by default which makes the labels seem oddly positioned. Here's a plot with the ticks on:
![figure_3](https://user-images.githubusercontent.com/302469/29853460-17dbd468-8d0e-11e7-80c6-cc0bfeeb650d.png)
I'm not totally sure how well parallel ticks work on a full-circle plot. It might be nicer to have the labels "above" the circle.

> This could be a useful option--but only if the upside-down labels are flipped 180 degrees to bring them right-side-up. In other words, label orientations should be kept within +-90 degrees, not +-180.

Easy to do, really, though I guess the question is about customization then and how easy that is to override. Especially tied into the next point.

> Just took a look at the rotatelabels feature of pie(), and it made me realize that there might be two ways people would want the labels oriented: perpendicular to the normal, or parallel to the normal.

We already have support for this in Cartesian plots: `ax.xaxis.set_tick_params(rotation=90)` and it mostly works here too except the anchor point is wrong (I can investigate that later.)

QuLogic commented:
To summarize, I'd like this on for wedges, but there already many special cases to preserve the old behaviour with full circle plots, so I'm fine with adding a few more to do the same for tick labels. I don't think there any point in preserving old ticks themselves because they were useless before.

dopplershift commented:
To be clear I think the new code with proper polar ticks is a nice upgrade. I just think the new alignment/rotation constitutes a style change that would not be universally well-received, so maintaining the old look (at least for full-circle plots) is the way to go.

dopplershift commented:
Also, I'm 👎 on holding up cutting the RC for this, so somebody had better get working on the modifications to make this behavior optional if it's going to land in the RC this week. 😈

anntzer commented:
2c from the peanut gallery: the rotated labels would probably look better if their "horizontal" (= in the axis of the text) center was tangent to the circle, not their "left" edge.

QuLogic commented:
It's straightforward to replicate old behaviour because there are already several checks for full-circle vs wedges. The only problem is that now that we are properly interpreting tick padding, the default is too small on figures of the size in the tests. Note though that since previously the angles were just offset by a fixed radial distance, any other sized figure might have had too much or too little padding with no way to fix it. I'm not really sure what to do about this.

QuLogic commented:
I'm not totally sure I like that hack to get padding to match previous tests. Maybe the best option is to add an rcParam for polar plots with a better default?

tacaswell commented:
On a bit more consideration, I am convinced that changing the default rotation is a style change which we have already determined is an API break so :+1: to these classes going in :-1: to rotating the tick labels by default on full circles :+1: to rotated by default on wedges (as those are new) and :+1: to a convenience method on the axes objects to apply the rotation.

I am 50/50 on letting this be merged post RC assuming it can be made to not break tests.

QuLogic commented:
Current results are 99% the same; it's very hard to get the padding exactly the same since it was never quite correct in the first place.

I keep forgetting that there are a couple polar tests that output multiple files, so that's why tests are failing right now.

tacaswell commented:
I am fine with minor padding shifts.

tacaswell commented:
https://github.com/matplotlib/matplotlib/pull/9068/files#diff-f1d864385683ce9e37a13df3975627c9 the labels have been squished up into the ticks.

QuLogic commented:
I figured out a simpler way to ensure tick transform independence (via `_copy_tick_props`) which cut down on some hacky `draw` overrides and fixed the axis labels.

I also modified some tests to enabled ticks on polar plots since those were disabled by default, and fixed a minor bug there as well.

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -240,6 +242,146 @@ def zoom(self, direction):
+        return t
+
+    def _apply_params(self, **kw):
+        maxis.XTick._apply_params(self, **kw)
```
Why not `super(ThetaTick, self)._apply_params(**kw)`?

QuLogic replied:
No reason; just couldn't remember which version was okay with old Python.

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -240,6 +242,146 @@ def zoom(self, direction):
+        self._text2_translate.invalidate()
+
+    def update_position(self, loc):
+        maxis.XTick.update_position(self, loc)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -240,6 +242,146 @@ def zoom(self, direction):
+        self.isDefault_majfmt = True
+
+    def cla(self):
+        maxis.XAxis.cla(self)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -240,6 +242,146 @@ def zoom(self, direction):
+        self._wrap_locator_formatter()
+
+    def _set_scale(self, value, **kwargs):
+        maxis.XAxis._set_scale(self, value, **kwargs)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -240,6 +242,146 @@ def zoom(self, direction):
+        'Copy the props from src tick to dest tick'
+        if src is None or dest is None:
+            return
+        maxis.XAxis._copy_tick_props(self, src, dest)
```
Super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+                    angle = self.axes._realViewLim.xmin
+                elif self.mode == 'max':
+                    angle = self.axes._realViewLim.xmax
+            angle %= 2 * np.pi
```
It only looks like `angle` is used below in trig functions--why is all of the playing with cycles of `2 * pi` necessary? They should just work, regardless, right?

QuLogic replied:
I recall something weird happening before; it might have been when things were mixed degrees/radians and I hadn't realized which units were in use at this location. I will double-check to see if this is necessary.

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+    preserve backwards compatibility.
+    """
+    def _get_text1(self):
+        t = maxis.YTick._get_text1(self)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+        return t
+
+    def _get_text2(self):
+        t = maxis.YTick._get_text2(self)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+        return t
+
+    def update_position(self, loc):
+        maxis.YTick.update_position(self, loc)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+        self.isDefault_majloc = True
+
+    def cla(self):
+        maxis.YAxis.cla(self)
```
super?

dopplershift commented on lib/matplotlib/projections/polar.py:
```diff
@@ -284,6 +426,181 @@ def view_limits(self, vmin, vmax):
+        self._wrap_locator_formatter()
+
+    def _set_scale(self, value, **kwargs):
+        maxis.YAxis._set_scale(self, value, **kwargs)
```
super?

efiring commented:
Like its predecessor, this is great progress, and I will be happy to see it go in ASAP--but I think it really needs some additional option controls, and a change to the default for wedges, before it goes in.  That could still be 2.1 post rc1, but since we are going to try very hard for a 2.2 a *short* time after 2.1, it might be better to target this to 2.2.

Specifically, at the very least, I think that the upside-down text as the default for wedges, illustrated in polar_theta_wedge.png, is not acceptable.  I could be wrong, but my suspicion is that the majority of people looking at a plot like that will be quite unhappy with it, and would be much happier with text rotated to be within +-90 degrees of horizontal.

I haven't looked closely at the code, or at all at the pie code, but I get the impression that there might be some aspects related to text positioning and rotation that could be factored out and thereby handled in a cleaner way.   Sorry that's so vague.

QuLogic commented:
I ran a similar plot through GMT (which is a common comparison in my field) and it also does the same thing. It is straightforward to implement this for the angular ticks:
![polar_theta_wedge](https://user-images.githubusercontent.com/302469/29948775-ca863f56-8e7e-11e7-9ab2-83a07fefb34d.png)
though the question is how to do configuration (if you set a custom rotation, maybe you don't want the automatic always-up mode?)

The other thing that GMT does is autorotate the radial labels as well. This is a bit harder to do since we need to set the anchor based on where the spine is going to be placed. I'm not sure if we can have that update automatically when you change the limits (it may already be automatic) or whether I'd need to throw in a bunch of `reset_ticks`...

QuLogic commented:
On second thought, the angle is already updated based on theta limits, so updating anchors won't be so difficult.

As for API, we already have `Axes.tick_params(rotate : Number)` and/or `Axis.set_tick_params(rotate : Number)` for setting the angle of tick labels. There are two possibilities:
1. Accept `rotate : [Number, String]`. If a number, use that exact rotation. If a string == `'auto'`, do the automatic stuff here (and use this as default for polar plots), and any other (future) options could be specified by string. Maybe also accept `Tuple(String, Number)` as well if we want to allow auto+additional rotation (to get perpendicular ones for example).
2. Add another argument `autorotate:Boolean` which triggers this PRs behaviour.

The former limits changes to Polar* classes only while the latter is perhaps a bit cleaner.

tacaswell commented:
I'm going to leave this un-merged for 2.1rc1, but if in the next two weeks it is ready :+1: to merging it for 2.1.

QuLogic commented:
There are a few edge cases to work out, but I've implemented what @efiring suggested locally. If we can decide on 1 or 2 for the configuration part, I think that deadline should be possible.

tacaswell commented:
I like option 1.

@QuLogic Any chance of this getting done today?

QuLogic commented:
Sure, give me a couple hours to make sure everything works and I can take care of it.

QuLogic commented:
I had rebased this earlier today and added the 'auto' option. Just now I made a small tweak so that 'auto' will also apply to full circles (since presumably that's what the user wants if they've explicitly asked for it) and modified one test to check that that works.

QuLogic commented on doc/users/whats_new.rst:
```diff
@@ -62,6 +62,13 @@ negative values are simply used as labels, and the real radius is shifted by
 the configured minimum. This release also allows negative radii to be used for
 grids and ticks, which were previously silently ignored.
 
+Radial ticks have be modified to be parallel to the circular grid line. Angular
```
Oops, small grammar typo here.

tacaswell commented:
backported to v2.1.x as 6889c6dbf

pmcculey commented:
if rotation of the theta labels is used in conjunction with ax.set_theta_zero_location( "N") and or ax.set_theta_direction( 1 ), the labels seem to rotate weirdly as you go around the theta axis. The following code should illustrate what I mean, the label flips through 180 degrees as you move from 90 to 135. I'm on a windows 10 PC & using python 2.7 btw.

import matplotlib.pyplot as plt
import numpy as np

fig = plt.figure()
rect = 0,0,1,1
ax = fig.add_axes(rect, label = 'ax', polar=True)
ax.patch.set_alpha(0)
ax.set_theta_zero_location( "N" )
ax.set_theta_direction( 1 )
thetaticks = np.arange( 0, 360, 45 )
ax.tick_params( 'x', pad = -15, rotation = 'auto' )

plt.show()

QuLogic commented:
That is fixed by #9881.
